### PR TITLE
Resolved parameter access problem in LocalDDP

### PR DESCRIPTION
**Pull request by ttochi**
### Issue
- If exec `param.free()` in model initialization, an error occurs in the backward process that the grad tensor size does not match.
- This issue occurs when wrapping the model with LocalDDP(DistributedDataParallel) which is directly accessing `param.data` that has already been offloaded.

### Resolve
- Remove the part that directly accesses `param.data` inside LocalDDP(DistributedDataParallel).
- Parameter fetch/free around the code block that hooking the gradient accumulation function
    - To obtain the gradient accumulation function by execute `param.expand_as(param)` in below code block, parameter should be fetch to gpu.
https://github.com/mcrl/Megatron-LM-mcrl/blob/67807825aa133b64d42e96b873d4632313f2ccec/megatron/model/distributed.py#L171-L180

### Discussion
- If fetch/free of modules is unavoidable during LocalDDP initialization, it may be better to modify `training.py` instead of `megatron/model/distributed.py`

### CODE REVIEW

File: megatron/model/distributed.py [hunk 1]
```diff
@@ -194,6 +203,7 @@ def zero_grad_buffer(self):
 
     def broadcast_params(self):
         for param in self.module.parameters():
+            # TODO (mcrl): check if spiral_tensor should be used instead of param.data
```
Comment (gajagajago) [inline megatron/model/distributed.py:206]: I think spiral_tensor is correct.

File: megatron/model/distributed.py [hunk 2]
```diff
@@ -145,22 +148,27 @@ def _get_buffer_type(param):
...
                     param.main_grad = self._grad_buffers[dtype].get(
-                        param.data.shape, type_num_elements[dtype])
+                        shape, type_num_elements[dtype])
                     if dtype not in self._grad_buffer_param_index_map:
                         self._grad_buffer_param_index_map[dtype] = {}
                     self._grad_buffer_param_index_map[dtype][param] = (
                         type_num_elements[dtype],
-                        type_num_elements[dtype] + param.data.nelement(),
+                        type_num_elements[dtype] + nelement,
                     )
```
Comment (gajagajago) [inline megatron/model/distributed.py:165]: For the above conditionals, please make on big conditional block for when `mpu.is_spiral_pipeline_parallel = True` and then set values according to parameter types (i.e., according to whether parameter is spiral_param or not). I think doing so improves readability and reduces future bug possibility.

File: megatron/training.py [hunk 3]
```diff
@@ -396,6 +396,7 @@ def get_model(model_provider_func, model_type=ModelType.encoder_or_decoder, wrap
         with ContextManagers(init_contexts):
             if args.DDP_impl == 'torch':
                 i = torch.cuda.current_device()
+                # NOTE (mcrl): For multiple microbatch, torchDDP is not supported.
                 model = [torchDDP(model_module, device_ids=[i], output_device=i,
```
Comment (gajagajago) [inline megatron/training.py:400]: I have never tested with torchDDP. Have you? Why is torchddp not supported for multi microbatch?
(If it is not crucial for us, we may for now prohibit using torch ddp with spiral by declaring so in the arguments.py check)

**Review by gajagajago — CHANGES_REQUESTED** [review]
Few miscellaneous changes requested!

File: megatron/model/distributed.py (hunk 1 above)
Comment (ttochi) [reply megatron/model/distributed.py]: This part is changed since I changed the code to fetch parameter before initialize LocalDDP.

File: megatron/model/distributed.py (hunk 2 above)
Comment (ttochi) [reply megatron/model/distributed.py]: This part is changed since I changed the code to fetch parameter before initialize LocalDDP.

File: megatron/training.py (hunk 3 above)
Comment (ttochi) [reply megatron/training.py]: I tested with torchDDP, but it is blocked by below code block.
https://github.com/mcrl/Megatron-LM-mcrl/blob/eaf017634f2b3118e724a28c07eabc6c7a15b26a/megatron/training.py#L503-L505

But I changed to fetch parameter before start initializing DDP module in modified code.

**Comment by ttochi** [comment]
@gajagajago Please review new commit, Thanks! 😃

File: megatron/training.py [hunk 4]
```diff
@@ -387,31 +387,41 @@ def get_model(model_provider_func, model_type=ModelType.encoder_or_decoder, wrap
 
     if wrap_with_ddp:
         init_contexts = []
+        wrapped_model = []
```
Comment (gajagajago) [inline megatron/training.py:390]: I don't think we really need this `wrapped_model` list. All below things can be done in original model[]. Let's not add another unnecessary variable. I will approve the PR after this is changed!

Comment (gajagajago) [reply megatron/training.py:390]: Please try with python list.map(lambda)! [ref](https://tykimos.github.io/2020/01/01/Python_Lambda_Map/) I think it will be better for readability

Comment (ttochi) [reply megatron/training.py]: Reflect in new commit :)

**Review by gajagajago — APPROVED** [review]
Thanks! I will merge the PR.
